File: src/utility/Logger.ts
/**
 * File Logger with Rotating Logs
 *
 * Maintains up to 8 log files:
 * - current.log: Active log file for the current session
 * - bot-1.log to bot-7.log: Previous 7 log sessions (rotated)
 *
 * Features:
 * - Real-time writes to current.log
 * - Automatic rotation on bot restart
 * - Keeps only the latest 8 sessions
 * - Timestamps for all log entries
 * - Color-coded console output (preserved in files)
 */

import * as fs from "node:fs";
import * as path from "node:path";

export enum LogLevel {
	DEBUG = 0,
	INFO = 1,
	WARN = 2,
	ERROR = 3,
}

export class Logger {
	private static instance: Logger;
	private logDir: string;
	private archiveDir: string;
	private currentLogPath: string;
	private writeStream: fs.WriteStream | null = null;
	private maxLogs = 8; // current.log + 7 archived logs
	private streamReady = false; // Track if stream is ready for writes
	private minLevel: LogLevel = LogLevel.INFO; // Default log level

	private constructor() {
		this.logDir = path.join(process.cwd(), "log");
		this.archiveDir = path.join(this.logDir, "archive");
		this.currentLogPath = path.join(this.logDir, "current.log");

		// Set log level from environment variable
		const envLogLevel = process.env.LOG_LEVEL?.toUpperCase();
		if (envLogLevel === "DEBUG") this.minLevel = LogLevel.DEBUG;
		else if (envLogLevel === "INFO") this.minLevel = LogLevel.INFO;
		else if (envLogLevel === "WARN") this.minLevel = LogLevel.WARN;
		else if (envLogLevel === "ERROR") this.minLevel = LogLevel.ERROR;

		this.ensureLogDirectory();
		this.rotateLogsOnStartup();
		this.initializeWriteStream();
	}

	static getInstance(): Logger {
		if (!Logger.instance) {
			Logger.instance = new Logger();
		}
		return Logger.instance;
	}

	/**
	 * Ensure log directory exists
	 */
	private ensureLogDirectory(): void {
		if (!fs.existsSync(this.logDir)) {
			fs.mkdirSync(this.logDir, { recursive: true });
		}
		if (!fs.existsSync(this.archiveDir)) {
			fs.mkdirSync(this.archiveDir, { recursive: true });
		}
	}

	/**
	 * Rotate logs on startup
	 *
	 * current.log → archive/bot-1.log
	 * archive/bot-1.log → archive/bot-2.log
	 * archive/bot-2.log → archive/bot-3.log
	 * ...
	 * archive/bot-6.log → archive/bot-7.log
	 * archive/bot-7.log → deleted
	 */
	private rotateLogsOnStartup(): void {
		// If current.log exists, rotate it
		if (fs.existsSync(this.currentLogPath)) {
			// Delete oldest log (archive/bot-7.log)
			const oldestLog = path.join(this.archiveDir, `bot-${this.maxLogs - 1}.log`);
			if (fs.existsSync(oldestLog)) {
				fs.unlinkSync(oldestLog);
			}

			// Rotate existing logs (bot-6.log → bot-7.log, bot-5.log → bot-6.log, etc.)
			for (let i = this.maxLogs - 2; i >= 1; i--) {
				const oldPath = path.join(this.archiveDir, `bot-${i}.log`);
				const newPath = path.join(this.archiveDir, `bot-${i + 1}.log`);
				if (fs.existsSync(oldPath)) {
					fs.renameSync(oldPath, newPath);
				}
			}

			// Rotate current.log → archive/bot-1.log
			const firstArchive = path.join(this.archiveDir, "bot-1.log");
			fs.renameSync(this.currentLogPath, firstArchive);
		}

		// Create new current.log with header
		const header = `


🔷 Daemira - New Session
Started: ${new Date().toISOString()}

`;
		fs.writeFileSync(this.currentLogPath, header);
	}

	/**
	 * Initialize write stream for current.log
	 */
	private initializeWriteStream(): void {
		this.writeStream = fs.createWriteStream(this.currentLogPath, {
			flags: "a", // Append mode
			encoding: "utf8",
			autoClose: false, // Keep stream open
		});

		// Mark stream as ready when opened
		this.writeStream.on("open", () => {
			this.streamReady = true;
		});

		// Handle stream errors
		this.writeStream.on("error", (error) => {
			// Use process.stderr.write directly to avoid recursion
			process.stderr.write(`🔸 FileLogger write stream error: ${error}\n`);
			this.streamReady = false;
		});

		// Stream is ready immediately after creation (synchronous open)
		// But we'll mark it as ready on the 'open' event for safety
		this.streamReady = true;
	}

	/**
	 * Write log entry to current.log with level filtering
	 */
	private _log(level: LogLevel, message: string): void {
		// Filter by log level
		if (level < this.minLevel) {
			return;
		}

		// Format: HH:MM:SS.mmm (compact, readable)
		const now = new Date();
		const hours = now.getHours().toString().padStart(2, "0");
		const minutes = now.getMinutes().toString().padStart(2, "0");
		const seconds = now.getSeconds().toString().padStart(2, "0");
		const millis = now.getMilliseconds().toString().padStart(3, "0");
		const timestamp = `${hours}:${minutes}:${seconds}.${millis}`;

		// Add level prefix
		const levelPrefix = ["[DEBUG]", "[INFO]", "[WARN]", "[ERROR]"][level];
		const logEntry = `[${timestamp}] ${levelPrefix} ${message}\n`;

		// Try to write via stream first (preferred method)
		if (this.writeStream && this.streamReady) {
			try {
				const written = this.writeStream.write(logEntry);

				// If write buffer is full, the write will be queued automatically
				// The stream will emit 'drain' when ready for more data
				if (!written) {
					// Buffer is full, but write is queued - stream will handle it
					// We don't need to do anything here as the write is already queued
				}
				return; // Successfully wrote via stream
			} catch (error) {
				// Stream write failed, fall through to fallback
			}
		}

		// Fallback: write directly to file if stream isn't ready or write failed
		// This ensures we don't lose messages during initialization or stream errors
		try {
			fs.appendFileSync(this.currentLogPath, logEntry, "utf8");
		} catch (error) {
			// Ignore fallback errors to avoid recursion
			// Message is lost, but at least we don't crash
		}
	}

	/**
	 * Log debug message (level 0)
	 */
	debug(message: string): void {
		this._log(LogLevel.DEBUG, message);
	}

	/**
	 * Log info message (level 1)
	 */
	info(message: string): void {
		this._log(LogLevel.INFO, message);
	}

	/**
	 * Log warning message (level 2)
	 */
	warn(message: string): void {
		this._log(LogLevel.WARN, message);
	}

	/**
	 * Log error message (level 3)
	 */
	error(message: string): void {
		this._log(LogLevel.ERROR, message);
	}

	/**
	 * Log message (alias for info - backward compatibility)
	 */
	log(message: string): void {
		this.info(message);
	}

	/**
	 * Write raw message without timestamp
	 */
	logRaw(message: string): void {
		if (!this.writeStream) {
			return;
		}

		const logEntry = `${message}\n`;

		// Write to file - use try/catch for synchronous errors
		try {
			this.writeStream.write(logEntry);
		} catch (error) {
			// Handle synchronous write errors
			process.stderr.write(`🔸 FileLogger write error: ${error}\n`);
		}
	}

	/**
	 * Get list of all log files (current + archived)
	 */
	getLogFiles(): string[] {
		const files: string[] = [];

		// Add current.log
		if (fs.existsSync(this.currentLogPath)) {
			files.push("current.log (ACTIVE)");
		}

		// Add archived logs
		for (let i = 1; i < this.maxLogs; i++) {
			const logPath = path.join(this.archiveDir, `bot-${i}.log`);
			if (fs.existsSync(logPath)) {
				const stats = fs.statSync(logPath);
				files.push(`archive/bot-${i}.log (${stats.size} bytes, ${stats.mtime.toISOString()})`);
			}
		}

		return files;
	}

	/**
	 * Close write stream on shutdown
	 */
	close(): void {
		if (this.writeStream) {
			this.writeStream.end();
			this.writeStream = null;
		}
	}
}

/**
 * Intercept console.log and write to file
 */
export function interceptConsole(): void {
	const logger = Logger.getInstance();

	const originalLog = console.log;
	const originalError = console.error;
	const originalWarn = console.warn;

	console.log = (...args: unknown[]) => {
		const message = args.map((arg) => String(arg)).join(" ");
		logger.log(message);
		originalLog.apply(console, args);
	};

	console.error = (...args: unknown[]) => {
		const message = args.map((arg) => String(arg)).join(" ");
		logger.error(message);
		originalError.apply(console, args);
	};

	console.warn = (...args: unknown[]) => {
		const message = args.map((arg) => String(arg)).join(" ");
		logger.warn(message);
		originalWarn.apply(console, args);
	};

	// Handle graceful shutdown
	process.on("SIGTERM", () => {
		logger.close();
	});

	process.on("SIGINT", () => {
		logger.close();
	});

	process.on("exit", () => {
		logger.close();
	});
}

File: src/utility/Logger.go
package utility

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// LogLevel represents the severity of a log message
type LogLevel int

const (
	DEBUG LogLevel = iota
	INFO
	WARN
	ERROR
)

func (l LogLevel) String() string {
	switch l {
	case DEBUG:
		return "DEBUG"
	case INFO:
		return "INFO"
	case WARN:
		return "WARN"
	case ERROR:
		return "ERROR"
	default:
		return "UNKNOWN"
	}
}

// Logger provides logging capabilities with file rotation
type Logger struct {
	level      LogLevel
	logDir     string
	currentLog *os.File
	mu         sync.Mutex
	mode       string // "file", "cli", "journal"
}

var (
	instance *Logger
	once     sync.Once
)

// GetLogger returns the singleton logger instance
func GetLogger() *Logger {
	once.Do(func() {
		instance = &Logger{
			level:  INFO,
			logDir: "log",
			mode:   "file",
		}
		instance.init()
	})
	return instance
}

// NewLogger creates a new logger with the specified mode
func NewLogger(mode string, level LogLevel) *Logger {
	logger := &Logger{
		level:  level,
		logDir: "log",
		mode:   mode,
	}
	if mode == "file" {
		logger.init()
	}
	return logger
}

// init initializes the logger and performs log rotation
func (l *Logger) init() {
	// Create log directory if it doesn't exist
	if err := os.MkdirAll(l.logDir, 0755); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create log directory: %v\n", err)
		return
	}

	// Rotate existing logs
	l.rotateLogs()

	// Open current log file
	currentLogPath := filepath.Join(l.logDir, "current.log")
	file, err := os.OpenFile(currentLogPath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to open log file: %v\n", err)
		return
	}

	l.currentLog = file
}

// rotateLogs rotates existing log files
func (l *Logger) rotateLogs() {
	archiveDir := filepath.Join(l.logDir, "archive")
	if err := os.MkdirAll(archiveDir, 0755); err != nil {
		return
	}

	currentLogPath := filepath.Join(l.logDir, "current.log")

	// Check if current.log exists
	if _, err := os.Stat(currentLogPath); err == nil {
		// Move bot-7.log to bot-8.log (delete bot-8.log)
		for i := 7; i >= 1; i-- {
			oldPath := filepath.Join(archiveDir, fmt.Sprintf("bot-%d.log", i))
			newPath := filepath.Join(archiveDir, fmt.Sprintf("bot-%d.log", i+1))

			if i == 7 {
				// Delete the oldest log
				os.Remove(newPath)
			}

			// Rename if file exists
			if _, err := os.Stat(oldPath); err == nil {
				os.Rename(oldPath, newPath)
			}
		}

		// Move current.log to bot-1.log
		os.Rename(currentLogPath, filepath.Join(archiveDir, "bot-1.log"))
	}
}

// log writes a log message
func (l *Logger) log(level LogLevel, format string, args ...interface{}) {
	if level < l.level {
		return
	}

	l.mu.Lock()
	defer l.mu.Unlock()

	timestamp := time.Now().Format("15:04:05.000")
	message := fmt.Sprintf(format, args...)
	logLine := fmt.Sprintf("[%s] [%s] %s\n", timestamp, level.String(), message)

	switch l.mode {
	case "file":
		if l.currentLog != nil {
			l.currentLog.WriteString(logLine)
		} else {
			fmt.Fprint(os.Stderr, logLine)
		}
	case "cli":
		l.printColoredLog(level, timestamp, message)
	case "journal":
		// For systemd journal, we'll use simple stdout
		fmt.Print(logLine)
	default:
		fmt.Print(logLine)
	}
}

// printColoredLog prints a colored log message to the console
func (l *Logger) printColoredLog(level LogLevel, timestamp, message string) {
	const (
		colorReset  = "\033[0m"
		colorBlue   = "\033[0;34m"
		colorGreen  = "\033[0;32m"
		colorYellow = "\033[1;33m"
		colorRed    = "\033[0;31m"
	)

	var color string
	switch level {
	case DEBUG:
		color = colorBlue
	case INFO:
		color = colorGreen
	case WARN:
		color = colorYellow
	case ERROR:
		color = colorRed
	default:
		color = colorReset
	}

	fmt.Printf("%s[%s] [%s]%s %s\n", color, timestamp, level.String(), colorReset, message)
}

// Debug logs a debug message
func (l *Logger) Debug(format string, args ...interface{}) {
	l.log(DEBUG, format, args...)
}

// Info logs an info message
func (l *Logger) Info(format string, args ...interface{}) {
	l.log(INFO, format, args...)
}

// Warn logs a warning message
func (l *Logger) Warn(format string, args ...interface{}) {
	l.log(WARN, format, args...)
}

// Error logs an error message
func (l *Logger) Error(format string, args ...interface{}) {
	l.log(ERROR, format, args...)
}

// Raw logs a message without timestamp or level
func (l *Logger) Raw(message string) {
	l.mu.Lock()
	defer l.mu.Unlock()

	if l.currentLog != nil {
		l.currentLog.WriteString(message + "\n")
	} else {
		fmt.Println(message)
	}
}

// SetLevel sets the minimum log level
func (l *Logger) SetLevel(level LogLevel) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.level = level
}

// Close closes the log file
func (l *Logger) Close() error {
	l.mu.Lock()
	defer l.mu.Unlock()

	if l.currentLog != nil {
		return l.currentLog.Close()
	}
	return nil
}

// GetWriter returns an io.Writer for the logger
func (l *Logger) GetWriter() io.Writer {
	if l.currentLog != nil {
		return l.currentLog
	}
	return os.Stdout
}

// ListLogFiles returns a list of all log files
func (l *Logger) ListLogFiles() []string {
	files := []string{}

	// Add current log
	currentLogPath := filepath.Join(l.logDir, "current.log")
	if _, err := os.Stat(currentLogPath); err == nil {
		files = append(files, currentLogPath)
	}

	// Add archived logs
	archiveDir := filepath.Join(l.logDir, "archive")
	if entries, err := os.ReadDir(archiveDir); err == nil {
		for _, entry := range entries {
			if !entry.IsDir() {
				files = append(files, filepath.Join(archiveDir, entry.Name()))
			}
		}
	}

	return files
}
